### DISCUSSION TITLE
Json objelerini Ruby'de modellemek

### DISCUSSION THREAD

**Discussion by BunyaminAkdeniz** [19/classroom#45] [Ruby] [answered]
Elimde basit bir json kaynağı var.  Response'lar şu şekilde geliyor. 
``` json
{
  "success": true,
  "result": [
    {
      "saat": "04:57",
      "vakit": "İmsak"
    },
    {
      "saat": "06:26",
      "vakit": "Güneş"
    },
    {
      "saat": "13:09",
      "vakit": "Öğle"
    },
    {
      "saat": "16:47",
      "vakit": "İkindi"
    },
    {
      "saat": "19:40",
      "vakit": "Akşam"
    },
    {
      "saat": "21:04",
      "vakit": "Yatsı"
    }
  ]
}
```
İnternette bazı kaynaklarda yakın örnekleirn openstruct ile yapıldığını gördüm ama siz openstructlardan uzak durmamızı söylemiştiniz. Zaten databasede null bir değer yok. Peki ben Ruby'de bunu nasıl modelleyebilirim.

**Comment by BunyaminAkdeniz** [comment]
Kodum şu şekilde 
```ruby
require 'uri'
require 'net/http'

url = URI("https://api.collectapi.com/pray/all?data.city=igdir")

http = Net::HTTP.new(url.host, url.port)
http.use_ssl = true
http.verify_mode = OpenSSL::SSL::VERIFY_NONE

request = Net::HTTP::Get.new(url)
request["content-type"] = 'application/json'
request["authorization"] = 'apikey ***************************************************'

response = http.request(request)

puts response.read_body #=>{"result":[{"saat":"03:16","vakit":"İmsak"},{"saat":"04:54","vakit":"Güneş"},{"saat":"12:06","vakit":"Öğle"},{"saat":"15:56","vakit":"İkindi"},{"saat":"19:08","vakit":"Akşam"},{"saat":"20:38","vakit":"Yatsı"}],"success":true}
```

**Comment by roktas** [comment, marked as answer]
Bu tip kodlar istenene bağlı olarak farklı şekillerde gerçeklenir.  API'den getirilen veriyi bir şekilde nesneye çevirirsin.  Ama önemli olan o nesnenin vazifesi veya büyük problemin çözümü içinde verinin hangi  nesne veya nesnelere dönüştürüleceği.  Dönüşüm sırasında validasyon da yapman gerekir.  Dış dünyadan aldığımız verilerde yani programı dış dünyaya bağlayan sınırlarda veri çekerken validasyon yapmamız gerekir.  Bu bazen kullanıcıdan alınan bir veri olabileceği gibi bu örnekte olduğu gibi bir dış servisten çekilen veri de olabilir.

"Bir şekilde nesneye çevirme" örneği şu olabilir.  API'den gelen "hash" önceden belirli olan anahtarlara sahip olduğundan en basiti `Struct` konteynerleri kullanmak. (OpenStruct türlü nedenlerle artık terkediliyor ama bu nedenleri bir yana bıraksan bile bu örnekte belirsiz sayıda anahtar olmadığından `Struct` doğru seçim, aksi halde `OpenStruct` gündeme alınabilirdi).

```ruby
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'

module API
  Vakit    = Struct.new :saat, :vakit, keyword_init: true
  Response = Struct.new :success, :result, keyword_init: true

  def self.call(response)
    data = JSON.parse(response, symbolize_names: true)
    Response.new success: data[:success], result: data[:result].map { |hash| Vakit.new(**hash) }
  end
end

p API.(DATA.read)

__END__
{
  "success": true,
  "result": [
    {
      "saat": "04:57",
      "vakit": "İmsak"
    },
    {
      "saat": "06:26",
      "vakit": "Güneş"
    },
    {
      "saat": "13:09",
      "vakit": "Öğle"
    },
    {
      "saat": "16:47",
      "vakit": "İkindi"
    },
    {
      "saat": "19:40",
      "vakit": "Akşam"
    },
    {
      "saat": "21:04",
      "vakit": "Yatsı"
    }
  ]
}
```

Daha production grade bir çözüm gerekiyorsa bunu net şekilde ortaya koyman lazım.  Bunu bilmediğimden validasyon düzenlemelerini de içeren basit bir gerçekleme örneği verebilirim ancak.

```ruby
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'set'
require 'time'

module API
  Response = Struct.new :success, :result, keyword_init: true

  def self.call(response)
    data = JSON.parse(response, symbolize_names: true)
    Response.new success: data[:success], result: data[:result].map { |hash| Vakit.parse(**hash) }
  end

  class Vakit
    VAKITLER = Set.new(%i[imsak sabah güneş öğle ikindi akşam yatsı]).freeze

    ParseError = Class.new StandardError

    attr_reader :saat, :vakit

    def initialize(saat:, vakit:)
      @saat  = saat
      @vakit = vakit
    end

    def self.parse(hash)
      begin
        saat  = Time.parse(hash[:saat])
      rescue ArgumentError => e
        raise ParseError, "Geçersiz namaz saati #{saat}: #{e.message}"
      end

      vakit = hash[:vakit].downcase(:turkic).to_sym
      raise ParseError, "Geçersiz namaz vakti #{vakit}" unless VAKITLER.include?(vakit)

      new saat: saat, vakit: vakit
    end
  end
end

p API.(DATA.read)

__END__
{
  "success": true,
  "result": [
    {
      "saat": "04:57",
      "vakit": "İmsak"
    },
    {
      "saat": "06:26",
      "vakit": "Güneş"
    },
    {
      "saat": "13:09",
      "vakit": "Öğle"
    },
    {
      "saat": "16:47",
      "vakit": "İkindi"
    },
    {
      "saat": "19:40",
      "vakit": "Akşam"
    },
    {
      "saat": "21:04",
      "vakit": "Yatsı"
    }
  ]
}
```

Bu örnekte problemin merkezinde `Vakit` adında bir nesne var, içini doldurmak sana kalmış.  `Vakit` nesnesini verinin sınır bölgesinde önce ayrıştırıyoruz (standart literatür).  Bu ayrıştırmada örneğin `saat` dizgisi (string) bir zaman nesnesine dönüşürken, `vakit` dizgisi de bir sembole dönüşüyor.  Elindeki probleme göre primitifler yerine bunların her biri ayrı birer özel nesne de olabilir.  Ayrıştırma sırasında validasyon da yapıyoruz.  Bu örnekte farketmen gereken kritik nokta ayrıştırma, validasyon vb işleri nesne düzeyinde değil sınıf düzeyinde metotlarla yapmak ve nesneyi ayrıştırılan bileşenlerden oluşan bir varlık halinde kendi başına bırakmak.  Neden böyle?  Çünkü ayrıştırma işlemi "construction"dan önce gelen bir işlem.  Bunu nesne içinde çözmeye çalışırsan ikilemle karşılaşırsın: contruct edeceğin nesneye ayrıştırma kodu koyarsan onu nasıl construct edeceksin?  Bir diğer yararı ise "test edilebilirlik".  Nesneleri veri alma biçimlerinden yalıtırsan daha etkin şekilde test edebilirsin.

P.S. Nesne adlarını API'den gelen verilere göre düzenledim.  Ben bu tür Türkçe adlardan oluşan API'leri pek sevmiyorum.  API'de kullanılan kelimeler gerçeklemede kullanacağın nesnelerdeki niteliklere bire bir olmasa bile yakın olmalı.  Yani API'ye baktığında problemin gerçeklemesinde rol oynayan aktörlere vereceğin uygun isimleri de kabaca görmen lazım.  Nedense Türkiye'de böyle kurgulanan çok API var.  Kullanıcıya bakan şeyler olsa anlayacağım ama bu API response'larının programatik değeri var; kullanıcıyı değil programcıyı ilgilendiren şeyler bunlar.  Tembel işi mi diyeyim, zevksizlik mi diyeyim bilemiyorum.  Maalesef ülkemizde çok yaygın bir pratik bu.  Muhtemelen kod yazarken de bu lexicon'u kullanıyorlar.  Yani `vakit` alanı varsa bir şekilde kodda da böyle geçiyor olmalı bu.  Sonuçta ortaya "Türkçe/İngilizce" karışık bir bulamaç çıkıyor.

Örnekleri Gist halinde alabilirsin: https://gist.github.com/roktas/f959fa7ac72617595ee20a5d3542a118
